### DISCUSSION TITLE
How to have dynamic menu placement based on the available viewport size?

### DISCUSSION THREAD

**Discussion by rfgamaral** [ariakit/ariakit#3479] [Questions] [answered]
Hi there,

To better demonstrate what I'm trying to achieve, I created the issue we are having with menu placement in this [CodeSandbox](https://codesandbox.io/p/sandbox/ariakit-menu-sample-vgc7rn).

By default, without setting `placement` in `Ariakit.MenuProvider`, we get this behaviour:

| Opening Menu `[1]` | Opening Menu `[2]` | Opening Menu `[3]` |
|--------|--------|--------|
| ![image](https://github.com/ariakit/ariakit/assets/96476/b36c24a2-8870-4c72-b213-e090ee231c5d) | ![image](https://github.com/ariakit/ariakit/assets/96476/d29c4235-ca79-43e4-ad5b-3c211983a992) | ![image](https://github.com/ariakit/ariakit/assets/96476/d40fbb9d-8e04-44e3-8bb2-b3e2dca4e6b9) |

The first menu is opened to the bottom, because that's the default. The second and third menus are opened to the top because they can't find enough space to render the **full menu** on the bottom (the default). However, for the **second menu**, there's also not enough space to render the **full menu** to the top, so why isn't this menu rendered to **the right**?

> [!NOTE]
> Passing a `placement` prop to the `SampleMenu` component is not a viable option to fix the issue, this is just a sample to demonstrate the problem we are having when using a component similar to `SampleMenu` in a list of dynamic sidebar items that goes from the top to the bottom of the viewport. In other words, we have no way fo knowing which ones need to have which placement.

So, is it possible to have the menu placement be completely dynamic based on the available viewport size, making sure to pick the best placement option to attempt to render the full menu at all times? Of course, if the viewport is too small on every side, that won't be possible, but that's okay.

**Comment by diegohaz** [comment, marked as answer]
By default, the [`flip`](https://ariakit.org/reference/menu#flip) feature only fallbacks to the opposite placement. You can provide a string with multiple fallback placements, separated by spaces:

```jsx
<Menu flip="top-start right">
```

**Reply by rfgamaral** [reply]
There are so many props for both `Menu` and `MenuProvider` (which is a good thing) that I must've missed that one from the documentation. Sorry about that.

**Reply by rfgamaral** [reply]
Sorry, @diegohaz, was just testing this, and with `flip="top right"`, it seems that the third menu now flips to the right side instead of the top (which has space for on my machine).

**Reply by diegohaz** [reply]
> Sorry, @diegohaz, was just testing this, and with `flip="top right"`, it seems that the third menu now flips to the right side instead of the top (which has space for on my machine).

Oh, the default value uses the same alignment, so it's `top-start right`.

**Reply by rfgamaral** [reply]
Perfect, @diegohaz. Thank you so much!
